common/scala/src/main/scala/whisk/core/containerpool/ContainerApi.scala

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.containerpool
+
+import java.time.Instant
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.Failure
+import scala.util.Success
+import spray.json.JsObject
+import spray.json.DefaultJsonProtocol._
+import whisk.common.Logging
+import whisk.common.LoggingMarkers
+import whisk.common.TransactionId
+import whisk.core.entity.ActivationResponse
+import whisk.core.entity.ActivationResponse.ContainerConnectionError
+import whisk.core.entity.ActivationResponse.ContainerResponse
+import whisk.core.entity.size._
+import whisk.http.Messages
+
+trait ContainerApi { container: Container =>
```

Comment:
Is this type narrowing useful? I thought about doing the same abstraction a while ago aswell. Would it make sense to instead model it as 

```scala
abstract class HttpContainer(id: ContainerId, ip: ContainerIp)(implicit ec: ExecutionContext, logging: Logging) extends Container
```

Reply:
Oh got it, this is the "formalized" version of something the runtime needs to implement and stuff. Hmm. Might be sensible. Thinking about it.

Reply:
e.g. `MesosTask extends Container with ContainerApi` - an alternative is to move the HTTP contract definition into Container?

Reply:
Right, we could just fold it into `Container` itself or have an intermediary type between `HttpContainer` which is exactly about the Http communication. Both sound sensible to me, cause I think all the container impls will just have `with ContainerApi`.